src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -486,46 +496,74 @@ private boolean checkSynchronized(final DetailAST synchronizedAst, boolean useBr
-            for (int i = endLineNo - 2; i > startLineNo - 1; i--) {
-                final int[] line = getLineCodePoints(i);
-                if (!CodePointUtil.isBlank(line)) {
-                    allThroughComment = matchesComment(reliefPattern, i + 1);
-                    break;
-                }
+            ast = currentCase;
+            lastAST = currentCase.getParent().getLastChild();
+        }
+        boolean reliefComment = false;
+        final List<DetailAST> commentContent = getCommentContentNodes(ast, lastAST);
+        for (DetailAST detailAST : commentContent) {
+            if (reliefPattern.matcher(detailAST.getText()).find()) {
+                reliefComment = true;
             }
         }
-        return allThroughComment;
+        return reliefComment || commentAsChildOfSameGroup(reliefPattern, currentCase);
     }
 
     /**
-     * Does a regular expression match on the given line and checks that a
-     * possible match is within a comment.
+     * This method will check if the comment are child of same case group ast.
      *
-     * @param pattern The regular expression pattern to use.
-     * @param lineNo The line number in the file.
-     * @return True if a match was found inside a comment.
+     * @param reliefPattern The regular expression pattern to match the comment text
+     * @param currentCase AST of the case that falls through to the next case.
+     * @return true if relief comment found
      */
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
-    private boolean matchesComment(Pattern pattern, int lineNo) {
-        final String line = getLine(lineNo - 1);
-
-        final Matcher matcher = pattern.matcher(line);
-        return matcher.find()
-                && getFileContents().hasIntersectionWithComment(
-                        lineNo, matcher.start(), lineNo, matcher.end());
+    private static boolean commentAsChildOfSameGroup(Pattern reliefPattern,
+                                                     DetailAST currentCase) {
+        boolean reliefComment = false;
+        final List<DetailAST> commentContent = getCommentContentNodes(currentCase.getLastChild(),
+                currentCase.getLastChild().getLastChild());
+        for (DetailAST detailAST : commentContent) {
+            if (reliefPattern.matcher(detailAST.getText()).find()) {
+                reliefComment = true;
+            }
+        }
+        return reliefComment;
     }
 
+    /**
+     * This method will return the comment content node between two nodes.
+     *
+     * @param fromNode starting node
+     * @param toNode last node to check
+     * @return list of commentContent node
+     */
+    private static List<DetailAST> getCommentContentNodes(DetailAST fromNode, DetailAST toNode) {
+        final List<DetailAST> commentContentNode = new LinkedList<>();
+        DetailAST node = fromNode;
+        while (true) {
```

Comment:
this infinite loop worries me. can we put condition of break in it ?

you kind of need to repeat logic of https://github.com/checkstyle/checkstyle/blob/4d6e3adc890b6d5abf362b85c9024f6771635027/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L403-L415
with extra code to collect notes of specific type and instead of until end (null) do until toNode.

Reply:
or https://github.com/sevntu-checkstyle/sevntu.checkstyle/blob/master/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/MapIterationInForEachLoopCheck.java#L756 we still do not use infine loop

Reply:
done